src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -183,126 +318,187 @@ number of sites `M` inferred from the number of modes in `address`.
-        x_sq = map(x -> Tuple(x).^2, CartesianIndices(ranges))
-        pot_vec = zeros(prod(S), C) # or undef...
+        ranges = Tuple(range(-fld(M, 2); length=M) for M in S)
+        x_sq = map(x -> Tuple(x) .^ 2, CartesianIndices(ranges))
+        pot_vec = zeros(prod(S), C)
         for c in 1:C
-            pot_vec[:,c] .= vec(map(x -> sum(v_mat[c,:] .* x), x_sq))
+            pot_vec[:, c] .= vec(map(x -> sum(v_mat[c, :] .* x), x_sq))
         end
     end
 
-    return HubbardRealSpace{C,typeof(address),typeof(geometry),D,typeof(t_vec),typeof(u_mat),typeof(v_mat),typeof(pot_vec)}(
-        address, t_vec, u_mat, v_mat, pot_vec, geometry,
+    return HubbardRealSpace{
+        TT,
+        C,
+        typeof(address),
+        typeof(geometry),
+        D,
+        typeof(t_mat),
+        typeof(u_mat),
+        typeof(w_mat),
+        typeof(v_mat),
+        typeof(pot_vec),
+    }(
+        address, t_mat, u_mat, w_mat, v_mat, pot_vec, geometry
     )
 end
 
+# Convert input of t or v to static matrix
+function _t_or_v_to_matrix(name, value, num_comps, num_dims; zero_is_nothing=true)
+    if zero_is_nothing && iszero(value)
+        return nothing
+    else
+        if value isa Number
+            value = fill(value, (num_comps, num_dims))
+        elseif size(value, 2) == 1 # column-vector
+            value = reduce(hcat, value for _ in 1:num_dims)
+        elseif size(value, 1) == 1 # row-vector
+            value = reduce(vcat, value for _ in 1:num_comps)
+        end
+        if (size(value, 1), size(value, 2)) ≠ (num_comps, num_dims)
+            throw(
+                ArgumentError(
+                    "`$name` must be a number, $num_comps × $num_dims matrix, or vector of length $num_comps",
+                ),
+            )
+        end
+        return SMatrix{num_comps,num_dims,float(eltype(value))}(value)
+    end
+end
+# Convert input of u or w to static matrix
+function _u_or_w_to_matrix(name, value, num_comps)
+    if iszero(value)
+        return nothing
+    else
+        if value isa Number
+            value = fill(value, (num_comps, num_comps))
+        end
+        if (size(value, 1), size(value, 2)) ≠ (num_comps, num_comps) ||
+            value isa Matrix && !issymmetric(value)
+            throw(
+                ArgumentError(
+                    "`$name` must be a number or a symmetric $num_comps × $num_comps matrix"
+                ),
+            )
+        end
+        return SMatrix{num_comps,num_comps,Float64}(value)
+    end
+end
+
 """
     warn_fermi_interaction(address, u)
 
-Warn if interaction matrix `u` does not make sense for `address`.
+Warn and modify if interaction matrix `u` does not make sense for `address`.
 """
 function warn_fermi_interaction(address::CompositeFS, u)
     C = num_components(address)
     for c in 1:C
-        if address.components[c] isa FermiFS && u ≠ ones(C,C) && u[c,c] ≠ 0
+        if address.components[c] isa FermiFS && u[c, c] ≠ 0
```

Comment:
There used to be a check here that didn't print the warning if `u` was at its default value. I don't think this should be printed when the user doesn't pass a value of `u` to the constructor.